simpleviewinc/keg-herkin
# ZEN-534: Configurable Mount

**Ticket**: [ZEN-534](https://jira.simpleviewtools.com/browse/zen-534)

## Context

* We need to setup up a configurable mounting environment 

## Goal

* Create a volume that is:
  * completely configurable by the consumer
  * has no utility files in it anymore
  * can still leverage our default cucumber step definitions inside consumer feature files
  * is compatible with qawolf work done previously

* Setup the herkin config file

## Updates

* `configs/aliases.config.js`
  * I set up module require aliases
  * To be compatible with jest, I also export `jestAliases` here
* `configs/getHerkinConfig.js`
  * getter for the herkin config
  * added toggleable warnings for missing custom config
  * if config is omitted from params, it now now searches the current working directory up to the root for that config
* `configs/herkin.default.config.js`
  * renamed from `app.config.js`
  * also renamed some of the paths, and ensured they first check the envs for values 
* `configs/jest-playwright.config.js`
  * configured with module aliasing
* `configs/jest.cucumber.config.js`
  * now dynamically searches for step definitions and support files
  * configured with module aliasing
* `configs/jest.e2e.config.js`
  * configured with module aliasing
* `configs/qawolf.config.js`
  * udpated to use the herkin config for paths
* `container/docker-compose.yml`
  * new volume mount using env whose value is derived from herkin.config
* `package.json`
  * updated to use the latest release of `cucumber-jest` that includes the expressions fix from last week
* `repos/backend/config.js`
  * updated with new name for app.config
* `repos/backend/endpoints/components.js`
  * deleted at request from Lance
* `repos/backend/index.js`
  * updated backend entry point to use module aliasing
* `repos/backend/libs/definitions/definitions.js`
  * updated with new mount paths
* `repos/backend/libs/features/features.js`
  * updated with new mount paths
* `repos/backend/libs/fileSys/testFiles.js`
  * updated with new mount paths
* `repos/backend/server.js`
  * updated to use herkin config
* `repos/example/herkin.config.js`
  * added herkin config to example client 
* `repos/example/tests/bdd/features/example.feature`
  * implemented a feature file that uses both some of our default step definitions and also one the example repo's
* `repos/tap/utils/getConfig.js`
  * updated to use default config
* `tests/*`
  * migrated a lot of files out
  * most were moved into a new repo called `repos/testUtils`
* `tasks/definitions/index.js`
  * updated with module aliasing
* `tasks/definitions/tap/start.js`
  * updated to set environment variables for herkin config paths
* `tasks/definitions/wolf/**`
  * updated with module aliasing
  * updated wolf command with alias to "waypoint"
* `tasks/utils/validation/validateConfig.js`
  * util for validating the herkin config
  * mainly checks that paths exist on file system


## Testing

### Setup
* `keg herkin && keg pr 12`
* `keg herkin start --config ./repos/example/herkin.config.js --from keg-herkin:zen-534-mount-updates`
  * OR: `cd` into `repos/example` then run `keg herkin start --from keg-herkin:zen-534-mount-updates`
* Ctrl-c or open another terminal

### Backend Endpoints
* `curl -s http://localhost:5005/features -w "\n"`
  * the features endpoint will now be dynamically looking for features in the volume mount's feature directory
  * verify that the response contains the **example** repo's feature file 
    * title should be: `Example Feature!!!`
* `curl -s http://localhost:5005/definitions -w "\n"`
  * the definitions endpoint will now be dynamically looking for steps in the volume mount's step def directory
  * verify that the response contains the example repo's step definitions AND our default step definitions
    * the example repo's step content is: `Given('I navigate to {word}', ...)`
* `curl -s http://localhost:5005/files/load?file=bdd/features/example.feature -w "\n"`
  * the files endpoint now looks inside the volume mount
  * verify the response status is 200 and the file content is correct

### Tests
* `keg herkin cucumber test`
  * Verify that it runs the cucumber tests successfully
  * the feature it runs should be located at `tests/bdd/features/example.feature`
* `keg herkin waypoint test`
  * Verify the qawolf test runs and passes
  * the test it runs should be located at `tests/waypoint/foo.test.js`
